### PR TITLE
Improved Java 17 support and Java runtime docs.

### PR DESCRIPTION
**Pull request by gianm**
Some work towards #12838.

1) Add a "Java runtime" doc page with information about supported
   Java versions, garbage collection, and strong encapsulation.

2) Update asm and equalsverifier to versions that support Java 17.

3) Add additional "--add-opens" lines to surefire configuration, so
   tests can pass successfully under Java 17.

4) Switch openjdk15 tests to openjdk17.

5) Update FrameFile to specifically mention Java runtime incompatibility
   as the cause of not being able to use Memory.map.

6) Update SegmentLoadDropHandler to log an error for Errors too, not
   just Exceptions. This is important because an IllegalAccessError is
   encountered when the correct "--add-opens" line is not provided,
   which would otherwise be silently ignored.

7) Update example configs to use druid.indexer.runner.javaOptsArray
   instead of druid.indexer.runner.javaOpts. (The latter is deprecated.)

8) Added a `run-java` command that sets up the proper "--add-opens"
   and "--add-exports" parameters for Java 17, and update out-of-box
   configs to include `druid.indexer.runner.javaCommand=bin/run-java`
   so this gets used.

### CODE REVIEW

**Comment by FrankChen021** [comment]
I used to submitted a PR #12333 to reflect our support of JDK 11. 
If JDK 17 is supported now, I think we also need to update the doc updated in that PR.

**Comment by gianm** [comment]
Ah, thanks for pointing that out. It's not quite supported yet, this PR is just a step. https://github.com/apache/druid/issues/12838 describes some more items we will need. We'll at least need to update DataSketches Memory.

**Comment by gianm** [comment]
In CI, I ran into a problem where `pull-deps` was run without the necessary exports and opens. So in the latest commit I added a `run-java` command that sets this stuff up, and am using it in three places:

1. At build time when `pull-deps` is run
2. At run time for `run-druid` and `run-zk`
3. In the default server configs, for `druid.indexer.runner.javaCommand`, so it gets used for tasks too.

**Review by paul-rogers — COMMENTED** [review]
The changes appear to be the result of actually trying Java 17 out with Druid and, as such, the changes are likely valid given that the result works. Reviewed the docs and other bits, all of which look great. LGTM (non-binding)

File: docs/operations/java.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+## Selecting a Java runtime
+
+Druid fully supports Java 8 and 11. The project team recommends Java 11.
+
+The project team does not recommend running with Java 17, because certain Druid functionality is not currently
+compatible with Java 17.
+
+There are many free and actively-supported Java runtime environments available, such as
+[Corretto](https://aws.amazon.com/corretto/) and [Zulu](https://www.azul.com/downloads/?package=jdk#download-openjdk).
```
Comment (xvrl) [inline docs/operations/java.md:36]: should we also mention Adoptium Temurin https://adoptium.net/temurin/releases/ here as one of the non-vendor-specific alternatives?

**Review by FrankChen021 — APPROVED** [review]
Wow, all CI have passed. I have taken a loot at most of changes and it LGTM.

One thing I'm not sure is if `pull-deps` works under JRE 17.

File: docs/operations/java.md (hunk 1 above)
Comment (gianm) [reply docs/operations/java.md:36]: Good idea. I wasn't aware of that one. Added it.

**Comment by gianm** [comment]
> One thing I'm not sure is if `pull-deps` works under JRE 17.

I did notice that it is run as part of the openjdk17 CI, because there was a problem in an earlier version of this patch. So it works at least well enough for that usage. I didn't try further usages.

File: examples/bin/run-java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# under the License.
+
+WHEREAMI="$(dirname "$0")"
+JAVA_BIN="$(source "$WHEREAMI"/java-util && get_java_bin_dir)/java"
+if [ -z "$JAVA_BIN" ]; then
+  >&2 echo "Could not find java - please run $WHEREAMI/verify-java to confirm it is installed."
+  exit 1
+fi
+
+JAVA_MAJOR="$("$JAVA_BIN" -version 2>&1 | grep ' version ' | head -n1 | sed -E 's/.*version "([^"]+)".*/\1/' | cut -d. -f1)"
```
Comment (xvrl) [inline examples/bin/run-java:27]: you can avoid the `grep ' version ' | head -n1 [...]` by adding `-n` to sed and printing the matching version.
With a little tweaking you can also avoid the `cut -d. -f1`

```suggestion
JAVA_MAJOR="$("$JAVA_BIN" -version 2>&1 | sed -n -E 's/.* version "([^."]*).*/\1/p')"
```

Comment (gianm) [reply examples/bin/run-java:27]: Nice. I took your suggestion.

**Pull request merged**